HW2/P5/physics.pyx

```diff
@@ -69,17 +73,37 @@ cdef void sub_update(FLOAT[:, ::1] XY,
     ############################################################
     # SUBPROBLEM 2: use the grid values to reduce the number of other
     # objects to check for collisions.
-    for j in range(i + 1, count):
-        XY2 = &(XY[j, 0])
-        V2 = &(V[j, 0])
-        if overlapping(XY1, XY2, R):
-            # SUBPROBLEM 4: Add locking
-            if not moving_apart(XY1, V1, XY2, V2):
-                collide(XY1, V1, XY2, V2)
-
-            # give a slight impulse to help separate them
-            for dim in range(2):
-                V2[dim] += eps * (XY2[dim] - XY1[dim])
+
+    # I will check the 5X5 object for collisions with the center of number i ball!
+
+    # Get i x coordinates, y coordinte and the grid size
+    i_grid_x = <UINT>(XY[i, 0] / grid_spacing)
+    i_grid_y = <UINT>(XY[i, 1] / grid_spacing)
+    grid_size = <UINT>( (1.0 / grid_spacing) + 1 )
+    
+    for j in range(i_grid_x - 2, i_grid_x + 3):               # 5X5 3 = 3 + 1
+        # Check the border position
+        if ( j < 0 ) and ( j > (grid_size - 1) ) :
```

Comment:
I think you actually want to check `if ( j < 0 ) or ( j > (grid_size - 1) )` here, instead of `and`.